tiny_cnn/core/nnp_backend.h

```diff
@@ -56,40 +58,40 @@ class nnp_backend : public math_backend {
       const nnp_convolution_kernel_transform_strategy kernel_tf_strategy =
           nnp_convolution_kernel_transform_strategy_reuse;
 
-      const cnn_size_t input_channels = this->in_.depth_;
-      const cnn_size_t output_channels = this->out_.depth_;
+      const cnn_size_t input_channels = params_->in.depth_;
+      const cnn_size_t output_channels = params_->out.depth_;
 
       const nnp_size input_size = {
-          static_cast<size_t>(this->in_.width_),
-          static_cast<size_t>(this->in_.height_)
+          static_cast<size_t>(params_->in.width_),
+          static_cast<size_t>(params_->in.height_)
       };
 
       const nnp_size kernel_size = {
-          static_cast<size_t>(this->weight_.width_),
-          static_cast<size_t>(this->weight_.height_)
+          static_cast<size_t>(params_->weight.width_),
+          static_cast<size_t>(params_->weight.height_)
       };
 
-      /*const nnp_padding padding = {.top = static_cast<size_t>(pad.cpu_data()[0]),
-                                   .right = static_cast<size_t>(pad.cpu_data()[1]),
-                                   .bottom = static_cast<size_t>(pad.cpu_data()[0]),
-                                   .left = static_cast<size_t>(pad.cpu_data()[1])};*/
-      const nnp_padding padding = { 0, 0, 0, 0 };
+      const nnp_padding padding = {
+          static_cast<size_t>(params_->in_padded.width_),
+          static_cast<size_t>(params_->in_padded.height_),
+          static_cast<size_t>(params_->in_padded.height_),
+          static_cast<size_t>(params_->in_padded.width_)
+      };
 
       const float* input_pointer =
-          reinterpret_cast<const float*>(
-              this->conv_layer_worker_storage_[index].prev_out_padded_);
+          reinterpret_cast<const float*>(&in_data[0]->at(0));
 
       const float* kernel_pointer =
-          reinterpret_cast<const float*>(&in_data[0]->at(0));
+          reinterpret_cast<const float*>(&in_data[1]->at(0));
 
       const float* bias =
-          reinterpret_cast<const float*>(&in_data[1]->at(0));
+          reinterpret_cast<const float*>(&in_data[2]->at(0));
 
```

Comment:
@edgarriba 
I would keep on hold the implementation of the Session, it could be useful to optimize some parallel runs of a same model like Tensorflow does, but I think it needs a redesign of model declaration to take advantages of sessions.
For the Device class, I'll start to add some discovery methods on derived classes (like to find number of devices for OCLDevice or the number of cores for CPUDevice). You can also implement some general method on the Base class (like, set the device ID). 
I see a Device instance will be passed to a backend, so we could, for example, limit the computation to a specific GPU or to a certain number of cores.
If some backend doesn't support device filtering (I'm not sure NNPACK does it) the device param will be ignored (with an alert)